avr-hal-generic/src/i2c.rs

```diff
@@ -210,14 +210,14 @@ macro_rules! impl_twi_i2c {
     ) => {$crate::paste::paste! {
         /// I2C Master
         $(#[$i2c_attr])*
-        pub struct $I2c<CLOCK: $crate::clock::Clock, M> {
+        pub struct [<$I2c Master>]<CLOCK: $crate::clock::Clock, M> {
```

Comment:
> Error:
>
> ```text
>    error[E0422]: cannot find struct, variant or union type `I2c` in this scope
>      --> chips/atmega328p-hal/src/lib.rs:31:20
>       |
>    31 |         pub struct I2c {
>       |                    ^^^ not found in this scope
> ```
>
> I don't understand something. I thought that the chips were defining the `I2c` struct and then passing it into the macro for expansion. The macro would then match against the struct elements, and then expand them in the macro expansion. If the chips defined the I2c struct, then this error would not be here. So where is the I2c struct defined? Here? Or in the macro?

Macro errors are often very confusing, be very careful about interpreting them as is.

The syntax inside the macro, that is, everything between `impl_twi_i2c!{` and `}` is _not_ interpreted by the rust compiler at all, so nothing is defined here.  The tokens are passed to the macro which then generates _something_ from them.  This result is the only thing the rust compiler sees.

Next, the `I2c` is a macro token that gets used in multiple places in the expansion.  What the error tells you is that somewhere the token is used but no type of that name is found.  That makes sense, because you've changed the macro to only generate a struct named `$I2c Master` and `$I2c Slave` (= `I2cMaster` & `I2cSlave`).  So you've probably forgot to change some reference to the name somewhere ...

For example, in line 241 and 271, you still reference a type `$I2c` instead of `[<$I2c Master>]`.  Maybe take another look if there's more places that need to be changed.

Reply:
Ah, that really helped. Thx!

Reply:
Pretty much.  One more step for completeness:

`atf2svd` generates -> `svd` files -> **`svdtools` patches** -> `svd2rust` generates

The supplied ATDF files are ... substandard to say the least and there's a lot of registers/fields that need to be fixed up before we can feed them to svd2rust.  That's all found in [`avr-device/path`](https://github.com/Rahix/avr-device/tree/master/patch) if you're curious.